### PR TITLE
Support case class in Diffy

### LINKED ISSUE #28: Support case classes in Diffy
We can derive `Diffy[T]` for case classes using shapeless or [magnolia](https://github.com/propensive/magnolia) similar to this: https://github.com/nevillelyh/shapeless-datatype/tree/neville/diff/core/src/main/scala/shapeless/datatype/diff.

Also refactor `Diffy[T]` for Avro/BigQuery/Protobuf, etc. to share the same interface.

### PR DESCRIPTION
**Pull request by nhanloukiala**
Resolve #28. Implementation details:

- Convert case class `T` into `Map[String, Any]`
- Nested case classes is converted into nested Maps
- Implement `diff` on `Map[String, Any]`

TODO : 
- Remove/migrate `ratatool-shapeless` module
- Update tests for `BigDiffy` to use `diffCaseClass` method

### CODE REVIEW

File: ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/CaseClassDiffy.scala [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    (left, right) match {
+      case (Some(l: Map[String, Any]), Some(r: Map[String, Any])) => diffMap(l, r, pref)
+      case (Some(l), Some(r)) => Seq(Delta(pref, l, r, delta(l, r)))
+    }
+  }
+}
+
+object CaseClassDiffy {
+  private def asMap[T](in : T)(implicit diff: MapEncoder[T]) = diff.toMap(in)
+}
```
Comment (idreeskhan) [inline ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/CaseClassDiffy.scala]: Nit, newline. Style should complain, or be fixed if it doesn't

File: ratatool-diffy/src/test/scala/com/spotify/ratatool/diffy/CaseClassDiffyTest.scala [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    val result = dBar.apply(b1, b2)
+
+    result should contain (Delta("b1", 1, 2, NumericDelta(1.0)))
+    result should contain (Delta("b2.f1", "foo1", "foo2", StringDelta(1.0)))
+    result should contain (Delta("b2.f2", 1, 1, NumericDelta(0.0)))
+    result should contain (Delta("b2.f3", 3,3, NumericDelta(0.0)))
+    result should contain (Delta("b2.f5", List("foo1").asJava, Vector("foo2").asJava, UnknownDelta))
+  }
+
+  "CaseClassDiffy" should "support ignore with exact match case" in {
```
Comment (idreeskhan) [inline ratatool-diffy/src/test/scala/com/spotify/ratatool/diffy/CaseClassDiffyTest.scala]: Can we also test that the other fields are correctly compared if we use ignore?

File: ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/BigDiffy.scala [hunk 3]
```diff
@@ -276,6 +276,13 @@ object BigDiffy extends Command {
                    diffy: TableRowDiffy): BigDiffy[TableRow] =
     diff(sc.bigQueryTable(lhs), sc.bigQueryTable(rhs), diffy, keyFn)
 
+  /** Diff two SCollection[T] **/
+  def diffCaseClass[T : ClassTag : MapEncoder](lhs: SCollection[T],
```
Comment (idreeskhan) [inline ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/BigDiffy.scala]: Can we move this and the shapeless deps all into `ratatoolShapeless` Would prefer not to force users to pull in Shapeless unless they need it.

File: build.sbt [hunk 4]
```diff
@@ -141,7 +142,8 @@ lazy val ratatoolDiffy = project
       "org.apache.beam" % "beam-runners-direct-java" % beamVersion,
       "org.apache.beam" % "beam-runners-google-cloud-dataflow-java" % beamVersion,
       "com.twitter" %% "algebird-core" % algebirdVersion,
-      "joda-time" % "joda-time" % jodaTimeVersion
+      "joda-time" % "joda-time" % jodaTimeVersion,
+      "com.chuusai" %% "shapeless" % shapelessVersion
```
Comment (idreeskhan) [inline build.sbt]: Can remove this dep now :)

File: ratatool-shapeless/src/main/scala/com/spotify/ratatool/shapeless/CaseClassDiffy.scala [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+  override def apply(x: T, y: T): Seq[Delta] = {
+    diff(Some(asMap(x)), Some(asMap(y)))
+      .filter(f => !hnilIgnore.exists(f.field.contains(_)))
+  }
+
+  private def hnilIgnore = ignore ++ Set(HNil.toString)
+
+  private def diff(left: Any, right: Any, pref: String = "")
+  : Seq[Delta] = {
```
Comment (idreeskhan) [inline ratatool-shapeless/src/main/scala/com/spotify/ratatool/shapeless/CaseClassDiffy.scala]: Nit, can this be moved onto the same line as function def?

File: ratatool-shapeless/src/main/scala/com/spotify/ratatool/shapeless/CaseClassDiffy.scala [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+  * */
+  implicit def optionEncoder[K <: Symbol, V](implicit witness: Witness.Aux[K]):
+  MapEncoder[FieldType[K, Option[V]]] = {
+    val name = witness.value.name
+    createEncoder { v =>
+      Map(name -> v.getOrElse(null.asInstanceOf[V]))
+    }
+  }
+
+  //scalastyle:off
```
Comment (idreeskhan) [inline ratatool-shapeless/src/main/scala/com/spotify/ratatool/shapeless/CaseClassDiffy.scala]: Missed this, should this be left in here? Also the following line looks a bit too long

File: ratatool-shapeless/src/main/scala/com/spotify/ratatool/shapeless/CaseClassDiffy.scala (hunk 5 above)
Comment (nhanloukiala) [reply ratatool-shapeless/src/main/scala/com/spotify/ratatool/shapeless/CaseClassDiffy.scala]: 👍

File: ratatool-shapeless/src/main/scala/com/spotify/ratatool/shapeless/CaseClassDiffy.scala (hunk 6 above)
Comment (nhanloukiala) [reply ratatool-shapeless/src/main/scala/com/spotify/ratatool/shapeless/CaseClassDiffy.scala]: I was thinking between disable line-length style checking and using type alias to shorten the code :

```
type ME[T] = MapEncoder[T]
  type FT[K, V] = FieldType[K, V]
  type Aux[H, R] = LabelledGeneric.Aux[H, R]

  implicit def hlistEncoder0[K <: Symbol, H, T <: HList](implicit hEnc: Lazy[ME[FT[K, H]]],
                                                         tEnc: Lazy[ME[T]]): ME[FT[K, H] :: T] = {
    createEncoder[FieldType[K, H] :: T]{ in =>
      hEnc.value.toMap(in.head) ++ tEnc.value.toMap(in.tail)
    }
  }

  implicit def hListEncoder1[K <: Symbol, H, T <: HList, R <: HList](implicit
                                                                      wit: Witness.Aux[K],
                                                                      gen: Aux[H, R],
                                                                      encoderH: Lazy[ME[R]],
                                                                      encoderT: Lazy[ME[T]]
                                                                     ): ME[FT[K, H] :: T] =
    createEncoder(in =>
      encoderT.value.toMap(in.tail) ++ Map(wit.value.name -> encoderH.value.toMap(gen.to(in.head))))
```

**Comment by nhanloukiala** [comment]
Re-indenting the lines works for me.

**Pull request merged**